File: disassembler.py
# Python
from argparse import ArgumentParser
import json
import logging
import struct

# External

# Local
from opcodes import Opcode

class UnknownInstructionError(Exception):
    pass

def get_instructions():
    try:
        with open('instructions.json') as f:
            file_data = json.load(f)
    except FileNotFoundError as e:
        logging.error(e)
        exit()

    if len(file_data) != 256:
        logging.error('Incomplete instruction set')
        exit()

    return file_data

INSTRUCTION_TABLE = get_instructions()

class Disassembler(object):
    def __init__(self, filename):
        try:
            with open(filename, 'rb') as f:
                self._data = f.read()
        except FileNotFoundError as e:
            logging.error(e)
            exit()

        self._index = 0
        self._end = len(self._data)
        self._digits = len(str(len(self._data)))

        logging.basicConfig(level=logging.WARNING, 
            filename='logs/disassembler.py.log', filemode='w')
        logging.getLogger('disassembler')

    def _log_message(self, index, mnem, register=None, address=None, 
            immediate=None, register_pair=None, reset=None):
        if register and immediate:
            return '{0:0{1}x} {2:<6} {3},#{4:#04x}'.format(index, self._digits, 
                mnem, register, immediate)
        elif register and address:
            return '{0:0{1}x} {2:<6} {3},${4:x}'.format(index, self._digits, 
                mnem.ljust(5), register, address)
        elif address:
            if mnem in ('OUT', 'IN'):
                return '{0:0{1}x} {2:<6} ${3:02x}'.format(index, self._digits, 
                    mnem, address)
            else:
                return '{0:0{1}x} {2:<6} ${3:04x}'.format(index, self._digits, 
                    mnem, address)
        elif register_pair:
            return '{0:0{1}x} {2:<6} {3},{4}'.format(index, self._digits, mnem, 
                register_pair[0], register_pair[1])
        elif register:
            return '{0:0{1}x} {2:<6} {3}'.format(index, self._digits, mnem, 
                register)
        elif immediate:
            return '{0:0{1}x} {2:<6} #{3:#04x}'.format(index, self._digits, 
                mnem, immediate)
        elif reset:
            return '{0:0{1}x} {2:<6} {3:x}'.format(index, self._digits, mnem, 
                reset)
        else:
            return '{0:0{1}x} {2:<6}'.format(index, self._digits, mnem)

    def _log(self, index, size, mnem, operand=None):
        token = mnem.split()[0]

        if size == 1:
            return '{0:0{1}x} {2}'.format(index, self._digits, mnem)
        elif size == 2:
            if token in ('out', 'in'):
                return '{0:0{1}x} {2}${3:02x}'.format(index, self._digits, 
                    mnem, operand)
            else:
                return '{0:0{1}x} {2}#{3:02x}'.format(index, self._digits, 
                    mnem, operand)
        elif size == 3:
            if token in ('lxi'):
                return '{0:0{1}x} {2}#{3:04x}'.format(index, self._digits, 
                    mnem, operand)
            else:
                return '{0:0{1}x} {2}${3:04x}'.format(index, self._digits, 
                    mnem, operand)

    def run(self):
        while self._index < self._end:
            byte = self._data[self._index]
            mnem = INSTRUCTION_TABLE[byte][0]
            size = INSTRUCTION_TABLE[byte][1]

            if size == 1:
                operand = None
            elif size == 2:
                operand = self._data[self._index + 1]
            elif size == 3:
                start = self._index + 1
                end = start + 2
                operand = struct.unpack('<H', self._data[start:end])[0]

            msg = self._log(self._index, size, mnem, operand=operand)

            self._index += size
            print(msg)
            logging.debug(msg)

if __name__ == '__main__':
    arg_parser = ArgumentParser()
    arg_parser.add_argument('filename', help='File to be disassembled')
    args = arg_parser.parse_args()
    
    d = Disassembler(args.filename)
    d.run()

File: disassembler.rb
# Python
require "json"
require "optparse"
require "logger"

LOGGER = Logger.new("logs/disassembler.rb.log")
LOGGER.level = Logger::WARN

def get_instructions
<<-DOC
  JSON array of arrays from file.
  Each instruction is represented as an array, like:
    ["mvi    c,",  2]

  where the first element is the mnemonic and the second is the size of the 
  the instruction in bytes.

  There must be a total of 256 instructions to be complete. The array indices 
  correspond to the opcodes of Intel 8080 (i.e. 0x00 through 0xff)
DOC
  begin
      f = File.read("instructions.json")
      file_data = JSON.parse(f)
  rescue Errno::ENOENT => e
      LOGGER.error(e)
      exit
  end

  unless file_data.length == 256
      LOGGER.error('Incomplete instruction set')
      exit
  end

  return file_data
end

INSTRUCTION_TABLE = get_instructions()

class Disassembler
  def initialize(filename)
    begin
      @data = File.binread(filename)
    rescue Errno::ENOENT => e
      LOGGER.error(e)
      exit
    end
    
    @index = 0
    @end = @data.length
    @digits = @end.to_s.length
  end

  def output(size, mnem, operand=None)
<<-DOC
    Output immediate operands prepended with the '#' character.
    Output address operands prepended with the '$' character.
DOC
    token = mnem.split()[0]

    case size
    when 1
      return "%04x %s" % [@index, mnem]
    when 2
      if ['out', 'in'].include? token
        return "%04x %s$%02x" % [@index, mnem, operand]
      else
        return "%04x %s#%02x" % [@index, mnem, operand]
      end
    when 3
      if ['lxi'].include? token
        return "%04x %s#%04x" % [@index, mnem, operand]
      else
        return "%04x %s$%04x" % [@index, mnem, operand]
      end
    end
  end

  def run()
    while @index < @end
      byte = @data[@index].ord
      mnem = INSTRUCTION_TABLE[byte][0]
      size = INSTRUCTION_TABLE[byte][1]

      case size
      when 1
        operand = nil
      when 2
        operand = @data[@index + 1].ord
      when 3
        start = @index + 1
        end_ = start + 2
        operand = @data[start..end_].unpack('<S_')[0]
      end

      msg = output(size, mnem, operand=operand)
      @index += size

      puts msg
      LOGGER.debug(msg)
    end
  end
end

def main
  options = {}
  OptionParser.new do |opts|
    opts.banner = "Usage: disassembler.rb [filename]"

    opts.on("-h", "--help", "Prints this help") do
      puts opts
      exit
    end
  end.parse!

  filename = ARGV[0]
  d = Disassembler.new(filename)
  d.run()
end

if __FILE__ == $0
  main()
end
